[tanwartushar] iP

tanwartushar opened the pull request:
# Donna

> "I'm Donna, [I know everything](https://youtu.be/Ehxl807mUBk?t=291)" 

Donna is here to help you streamline your task management with ease :)

### What Donna Can Do 

- Add  `ToDo`, `Deadline`, or `Event` tasks
- Delete tasks
- List all the tasks
- Mark tasks as complete (or incomplete)
- Save tasks locally (and pick up from where you left!)
<sub>what can she not do?</sub>

**To get Donna's assistance with anything, follow these steps:** 

1. Start the chatbot
2. Enter the command
3. Let Donna handle the rest! :wink:

Features to be added:

- [x] Add a basic GUI
- [ ] Tweak the UI 

##### PS: This is how Donna marks tasks

```
Task task = tasks.markTask(taskIdx, false);
ui.printTaskMarkedMessage(task, false);
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
class renamed, greeting added

Commit:
echo added

Commit:
add, list commands added

Commit:
Task class added

Commit:
mark, unmark added

Commit:
ToDo class added

Commit:
Task updated; Deadline, Event class added

Commit:
Task toString updated

Commit:
support added for todo, event and deadline tasks

Commit:
DonnaException class added

Commit:
error handling

Commit:
deletion support added

Commit:
text UI testing

Commit:
feat: add file saving support

Commit:
feat: add file loading support

Commit:
Merge branch 'branch-Level-7'

Commit:
Deadline.java: add support for Date&Time

Commit:
Event.java: add support for Date&Time

Commit:
Merge branch 'branch-Level-8'

Commit:
Add Storage.java

Commit:
Add TaskList.java

Commit:
Add Ui.java

Commit:
Add Parser.java

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Refactor structure: Organise into packages

Commit:
Merge branch 'branch-A-Packages'

Commit:
Merge branch 'add-gradle-support'

Commit:
Add JUnit testing

Commit:
Update build.gradle

Commit:
Add support for find

Commit:
Deadline.java: Add Javadocs

Commit:
ToDo.java: Add Javadocs

Commit:
Event.java: Add Javadocs

Commit:
Storage.java: Add Javadoc

Commit:
ParsedCommand.java: Add Javadoc

Commit:
DonnaException.java: Fix indentation

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Donna.java: Update Javadocs

Commit:
Parser.java: Update Javadocs

Commit:
Storage.java: Update Javadocs

Commit:
TaskList.java: Update Javadocs

Commit:
Ui.java: Update Javadocs

Commit:
Deadline.java: Update Javadocs

Commit:
Event.java: Update Javadocs

Commit:
ToDo.java: Update Javadocs

Commit:
Task.java: Update Javadocs

Commit:
Add GUI using javaFX

Commit:
Refactor code to package classes

rachelyeohm commented on src/main/java/donna/Donna.java:
```diff
@@ -0,0 +1,180 @@
+        while (true) {
+            String input = sc.nextLine();
+            try {
+                ParsedCommand result = parser.parse(input);
```
If you don't need the parser class to be unique to each instance, it could be better to make it static as it's for utility.

tanwartushar replied:
Makes sense. I'll work on it

rachelyeohm commented on src/main/java/donna/Donna.java:
```diff
@@ -0,0 +1,180 @@
+                    ui.printTaskList(tasks);
+                    break;
+                case "mark":
+                    handleMark(result.getArgument1());
```
I really like that you separate your command into arguments and command name in the ParsedCommand - it's very neat.

tanwartushar replied:
Thank you!

rachelyeohm commented on src/main/java/donna/Donna.java:
```diff
@@ -0,0 +1,180 @@
+     * @param taskNum Index of the task to unmark (index from 1).
+     * @throws DonnaException If the argument is not a valid task number.
+     */
+    private void handleUnmark(String taskNum) throws DonnaException {
```
Since handleUnmark, handleAdd etc are all the same type of methods, it might be better to put them in a separate class of their own, or have different types of Commands and ways of executing them.

tanwartushar replied:
That's a valid point. I'll do so.

rachelyeohm commented on src/main/java/donna/Storage.java:
```diff
@@ -0,0 +1,107 @@
+
+            while (sc.hasNext()) {
+                String line = sc.nextLine();
+                String[] inWords = line.split(" \\| ");
```
I think this could be named more meaningfully / clearly :(

tanwartushar replied:
True! Will do.

rachelyeohm commented on src/main/java/donna/Ui.java:
```diff
@@ -0,0 +1,152 @@
+        System.out.println("| | | |/ _ \\| '_ \\| '_ \\ / _` |");
+        System.out.println("| |_| | (_) | | | | | | | (_| |");
+        System.out.println("|____/ \\___/|_| |_|_| |_|\\__,_|");
+        System.out.println();
```
so cute haha

rachelyeohm commented on src/main/java/donna/Ui.java:
```diff
@@ -0,0 +1,152 @@
+                System.out.println((i + 1) + ". " + tasks.get(i));
+            }
+        }
+        printDashedLine();
```
I would suggest, if you have to keep writing printDashedLine() before and after each paragraph of text, it would be better to have a special method that does it for you (to avoid repetition)

tanwartushar replied:
Thank you for pointing that out. Fixed it :)

rachelyeohm commented on src/test/java/donna/task/TaskTest.java:
```diff
@@ -0,0 +1,38 @@
+public class TaskTest {
+
+    @Test
+    void testTaskInitialisationWithValidDescription() throws DonnaException {
```
If you're feeling fancy, you could split the test method names into three parts like in the tutorial

tanwartushar replied:
Noted xD

Commit:
Update GUI

Commit:
Merge branch 'branch-Level-10'

ty-3401 commented on src/main/java/donna/Donna.java:
```diff
@@ -0,0 +1,228 @@
+
+import donna.parse.ParsedCommand;
+import donna.parse.Parser;
+import donna.task.*;
```
I think the tasks import should be listed explicitly.
https://se-education.org/guides/conventions/java/intermediate.html#statements

tanwartushar replied:
yes 😅 will fix it.

ty-3401 commented on src/main/java/donna/DonnaException.java:
```diff
@@ -0,0 +1,34 @@
+        super(message);
+    }
+
+    public static DonnaException emptyDescription(String taskType) {
+        return new DonnaException("Please provide the description for a " + taskType + " task");
+    }
+
+    public static DonnaException emptyDeadline() {
+        return new DonnaException("Please provide a deadline for this task!" + "\n"
+                + "Use /by to provide a deadline.");
+    }
+
+    public static DonnaException emptyEventTime() {
+        return new DonnaException("Please provide the timings for this event!"
+                + "\n" + "Use /from and /to to provide timings.");
+    }
+
+    public static DonnaException invalidTaskType(String taskType) {
+        return new DonnaException("I'm sorry I don't understand that :(\n"
+                + "Please use a valid task type \n"
+                + "(todo / deadline / event)\n"
+                + "or you may type list , delete or mark / unmark\n"
+                + "to view, delete or mark tasks");
+    }
+
+    public static DonnaException invalidTaskNumber() {
+        return new DonnaException("No task assigned to this number yet.\n"
+                + "Retry with a valid task number!");
+    }
```
I like how descriptive each of the error messages are! Kudos to you!

ty-3401 commented on src/main/java/donna/Storage.java:
```diff
@@ -0,0 +1,104 @@
+package donna;
+
+import donna.task.*;
```
Same as a previous comment. Imports should be listed explicitly.
https://se-education.org/guides/conventions/java/intermediate.html#statements

tanwartushar replied:
will fix it.

ty-3401 commented on src/main/java/donna/Ui.java:
```diff
@@ -0,0 +1,173 @@
+     */
+    public String getGreeting() {
+        StringBuilder greeting = new StringBuilder();
+//        greeting.append(donnaLogo());
```
I think if you're not gonna use this you can delete it. Otherwise I think it should be in line with the rest of the code.
https://se-education.org/guides/conventions/java/intermediate.html#comments

tanwartushar replied:
You're right. I left it for the text UI, but I can remove it now.

ty-3401 commented on src/main/java/donna/Ui.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    static private String donnaLogo() {
+        return " __      \n"
+                + "|  _  \\  __  _ __   _ _   __  \n"
+                + "|  |  |  | / _ \\| '_  \\ | '_ \\ / _` |\n"
+                + "|  |_| |  (_) |  | |  |  |  |   | (_|  |\n"
+                + "|___/ \\__/|_| |_|_|  |_|\\_,_|\n";
```
I love the ASCII art. I've also done the same in my code.

ty-3401 reviewed (commented):
Sorry for the late review! I hope you find them useful. I don't believe there is much I can find wrong with your code other than some nits.

tanwartushar commented:
> Sorry for the late review! I hope you find them useful. I don't believe there is much I can find wrong with your code other than some nits.

Thank you for the review! Will do the needful.

Commit:
Add CheckStyle

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Update build.gradle

Commit:
Refactor classes to use java assertions

Several important assumptions are undocumented, leaving the code
vulnerable to unexpected issues.

Use assert feature across the classes to define assumptions about the
program state.

Adding assertions help to verify assumptions about certain invariants
and detect and correct programming errors.

Commit:
Refactor Donna.java to improve code quality

Handle methods (handleAdd(), handleDelete(), etc.) have similar
functions.

These common functions are extracted to a new CommandHandler class.

Minor tweaks to improve code quality and readability.

Commit:
Update TaskList.java to use streams

tanwartushar committed:
Merge pull request #1 from tanwartushar/branch-A-Assertions

branch-A-Assertions

Commit:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Merge branch 'master' into branch-A-Streams

tanwartushar committed:
Merge pull request #2 from tanwartushar/branch-A-CodeQuality

branch-A-CodeQuality

Commit:
Merge branch 'master' into branch-A-Streams

tanwartushar committed:
Merge pull request #3 from tanwartushar/branch-A-Streams

branch-A-Streams

Commit:
Add support for tagging

Commit:
Add support for local storage of tags

Commit:
Add support for finding tasks by tag

Commit:
Merge branch 'branch-BCD-Extension'

Commit:
Update chatbot's responses

Commit:
Merge branch 'branch-A-Personality'

Commit:
Update gui

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Add user guide

Commit:
Update user guide

Commit:
Update JUnit test cases

Commit:
Update build.gradle
